src/titan_db_test.cc

```diff
@@ -1325,6 +1345,460 @@ TEST_F(TitanDBTest, GCAfterReopen) {
+  CompactAll(cfa);
+  CompactAll(cfa);
+
+  CheckBlobFileCount(0, cfa);
+}
+
+TEST_F(TitanDBTest, DeleteAllGC) {
+  options_.max_background_gc = 2;
+  options_.disable_background_gc = false;
+  Open();
+
+  ASSERT_OK(db_->Put(WriteOptions(), "k1", std::string(100, 'v')));
+  ASSERT_OK(db_->Put(WriteOptions(), "k2", std::string(10 * 1024, 'v')));
+  Flush();
+  ASSERT_OK(db_->Delete(WriteOptions(), "k1"));
+  ASSERT_OK(db_->Delete(WriteOptions(), "k2"));
+  Flush();
+  CompactAll();
+  CompactAll();
+
+  CheckBlobFileCount(0);
+}
+
+TEST_F(TitanDBTest, LowDiscardableRatio) {
+  options_.max_background_gc = 2;
+  options_.disable_background_gc = false;
+  options_.blob_file_discardable_ratio = 0.01;
+  Open();
+
+  ASSERT_OK(db_->Put(WriteOptions(), "k1", std::string(10 * 1024, 'v')));
+  auto snap = db_->GetSnapshot();
+  ASSERT_OK(db_->Put(WriteOptions(), "k1", std::string(100 * 1024, 'v')));
+  Flush();
+
+  db_->ReleaseSnapshot(snap);
+  CheckBlobFileCount(1);
+  std::shared_ptr<BlobStorage> blob_storage = GetBlobStorage().lock();
+  ASSERT_TRUE(blob_storage != nullptr);
+  std::map<uint64_t, std::weak_ptr<BlobFileMeta>> blob_files;
+  blob_storage->ExportBlobFiles(blob_files);
+  auto prev_file_size = blob_files.begin()->second.lock()->file_size();
+  CompactAll();
+  CompactAll();
+
+  CheckBlobFileCount(1);
+  blob_storage->ExportBlobFiles(blob_files);
+  ASSERT_TRUE(blob_files.begin()->second.lock()->file_size() < prev_file_size);
+}
+
+TEST_F(TitanDBTest, CompactionDuringGC) {
+  options_.max_background_gc = 2;
+  options_.disable_background_gc = false;
+  options_.blob_file_discardable_ratio = 0.01;
+  Open();
+
+  ASSERT_OK(db_->Put(WriteOptions(), "k1", std::string(10 * 1024, 'v')));
+  auto snap = db_->GetSnapshot();
+  ASSERT_OK(db_->Put(WriteOptions(), "k1", std::string(100 * 1024, 'v')));
+  Flush();
+
+  db_->ReleaseSnapshot(snap);
+  CheckBlobFileCount(1);
+  std::shared_ptr<BlobStorage> blob_storage = GetBlobStorage().lock();
+  ASSERT_TRUE(blob_storage != nullptr);
+  std::map<uint64_t, std::weak_ptr<BlobFileMeta>> blob_files;
+  blob_storage->ExportBlobFiles(blob_files);
+  auto prev_file_size = blob_files.begin()->second.lock()->file_size();
+  CompactAll();
+
+  SyncPoint::GetInstance()->LoadDependency(
+      {{"TitanDBTest::CompactionDuringGC::ContinueGC",
+        "BlobGCJob::Finish::BeforeRewriteValidKeyToLSM"},
+       {"BlobGCJob::Finish::AfterRewriteValidKeyToLSM",
+        "TitanDBTest::CompactionDuringGC::WaitGC"}});
+  SyncPoint::GetInstance()->EnableProcessing();
+
+  CompactAll();
+  ASSERT_OK(db_->Delete(WriteOptions(), "k1"));
+  blob_storage->ExportBlobFiles(blob_files);
+  ASSERT_EQ(blob_files.size(), 2);
+
+  TEST_SYNC_POINT("TitanDBTest::CompactionDuringGC::ContinueGC");
+  TEST_SYNC_POINT("TitanDBTest::CompactionDuringGC::WaitGC");
+
+  std::string value;
+  Status status = db_->Get(ReadOptions(), "k1", &value);
+  ASSERT_EQ(status, Status::NotFound());
+  CheckBlobFileCount(1);
```

Comment:
(just a question)

I'm wondering why  the blob file count is 1 instead of 0?
even if we do Compaction before CheckBlobFileCount here, the blob file is still not deleted.
```c++
  std::string value;
  Status status = db_->Get(ReadOptions(), "k1", &value);
  ASSERT_EQ(status, Status::NotFound());
  // ??? even if we do Compaction here, the blob file will not be deleted.
  Flush();
  CompactAll();
  CompactAll();
  CheckBlobFileCount(1);
  blob_storage->ExportBlobFiles(blob_files);
  ASSERT_TRUE(blob_files.begin()->second.lock()->file_size() < prev_file_size);
  // after the reopen, the GC will delete the blob file
  Reopen();
  CheckBlobFileCount(0);
```

Reply:
I'm investigating it

Reply:
This is the minimal reproduce code I found.
```c++
TEST_F(TitanDBTest, CompactionAndGC) {
  options_.max_background_gc = 1;
  options_.disable_background_gc = false;
  options_.blob_file_discardable_ratio = 0.01;
  Open();

  ASSERT_OK(db_->Put(WriteOptions(), "k1", std::string(10 * 1024, 'v')));
  auto snap = db_->GetSnapshot();
  ASSERT_OK(db_->Put(WriteOptions(), "k1", std::string(100 * 1024, 'v')));
  Flush();

  db_->ReleaseSnapshot(snap);
  CheckBlobFileCount(1);

  CompactAll();
  CompactAll();

  // wait till GC finished
  // if we don't wait here, the case will pass
  // Env::Default()->SleepForMicroseconds(1000 * 1000);

  ASSERT_OK(db_->Delete(WriteOptions(), "k1"));
  Flush();
  CompactAll();
  CompactAll();
  CheckBlobFileCount(0);
}
```